Configure what is being printed to stdout when executing command

HaxScramper opened the pull request:
I added compile-time and runtime configuration options for conditionally disable printing of the command's output or command itself.

+ New compile-time pragmas: `shellNoDebugCommand`, `shellNoDebugOutput` that can disable output for command. 
+ New proc `setShellDebugConfig` for setting shell debug configuration at runtime. 
+ New enum `DebugOutputKind` for controlling what is being printed at runtime

----

I also want to add support for getting `stderr` and `stdout` *separately* instead of merging them into one string. I don't want to implement it in a way that would break existing code using this module (if I were to add third field to resulting tuple it would certainly break things). If you have any ideas I would be happy to use them as base for my implementation.

I haven't made any changes in README because I first want to implement several other things (support for separate `stderr` handling is one of them).

Vindaar commented:
First of all thanks a lot! I'll take a proper look tomorrow.

Vindaar commented on shell.nim:
```diff
@@ -197,7 +227,15 @@ proc concatCmds(cmds: seq[string], sep = " && "): string =
+
+# NOTE Checking for key in set, calling function on each `asgnShell`
+# invokation might decrease performance. Although not printing
+# anything into stdout might increase performace.
```
I'm pretty sure that compared to the overhead of invoking a shell in the first place a simple check of "flag in set" is negligible.

Vindaar commented on shell.nim:
```diff
@@ -220,9 +259,26 @@ proc asgnShell*(cmd: string): tuple[output: string, exitCode: int] =
+      # might be possible to either change function signature (not
+      # desired as it breaks API) or provide optional parameter for
+      # function call (`var errorRes: Option[string]` or something
+      # like that)
```
I wouldn't worry too much about breaking the API tbh. `asgnShell` wasn't intended as a user facing proc in the first place. It's mainly exported for the tests, iirc. 

And if anyone used `asgnShell` manually, that's what versioning is for (I'd release version v0.3.0 with your PR for that reason). So feel free to return 
```nim
tuple[output, error: string, exitCode: int]
```
instead.

Vindaar commented on shell.nim:
```diff
@@ -327,13 +389,17 @@ macro shellVerbose*(cmds: untyped): untyped =
   for cmd in shCmds:
     let qCmd = nilOrQuote(cmd)
     result.add quote do:
+      let debugConfig = `debugConfig`
```
This shouldn't be necessary. Inserting `debugConfig` below should be fine.

Vindaar commented on shell.nim:
```diff
@@ -14,6 +14,36 @@ type
+  dokRuntime
+}
+
+var globalConfigChanged: bool = false
```
I'm not a big fan of global variables like this. I understand you include it to support setting flags at runtime, but I feel like we should be able to achieve the same without globals.
Especially, because this means that all `shell` invocations will be set to the same flags. 

The `debugConfig` given to `shellVerbose` now already allows to give a variable, which can be set at runtime, no? Of course something like your example below:
```nim
  let (res, _) = shellVerbose:
￼    echo "test"
￼
￼  echo "Result is: ", res
￼
￼  setShellDebugConfig({})
￼
￼  let (res2, _) = shellVerbose:
￼    echo "test2"
```
has to be something like:
```nim
  var myConfig: set[DebugOutputKind] = someConfig()
  let (res, _) = shellVerbose(myConfig):
￼    echo "test"
￼
￼  echo "Result is: ", res
￼
￼  myConfig.clear()

￼  let (res2, _) = shellVerbose(myConfig):
￼    echo "test2"
```
But I prefer that. Global state like this can be *really* confusing. Plus in that case we don't even need the `setShellDebugConfig` and `getShellDebugConfig` procs at all. Nim's syntax for `enum` and `set` is clearer imo, too.

Vindaar commented on shell.nim:
```diff
@@ -409,3 +480,22 @@ macro shellAssign*(cmd: untyped): untyped =
+    shellVerbose {dokOutput}:
+      echo "Result"
+
+  echo "Result 3 is:", res3
```
These should become proper tests.

Vindaar commented on shell.nimble:
```diff
@@ -1,6 +1,6 @@
 # Package
 
-version       = "0.2.3"
+version       = "0.2.4"
```
Make that `v0.3.0`.

HaxScramper commented:
I removed global state, cleaned up comments and added support for returning stderr result. I decided it would be a good idea to leave `shellVerbose` the way it works now and instead defined another macro `shellVerboseErr` that returns `stderr` and `stdout` separately.

Vindaar commented on shell.nim:
```diff
@@ -220,31 +251,65 @@ proc asgnShell*(cmd: string): tuple[output: string, exitCode: int] =
+
     let exitCode = pid.peekExitCode
+
+    # Zero exit code does not guarantee that there will be nothing in
```
Heh, good point. Shows my use case so far, haha

Vindaar commented on shell.nim:
```diff
@@ -220,31 +251,65 @@ proc asgnShell*(cmd: string): tuple[output: string, exitCode: int] =
+
+  # IDEA split command by space and layout it line-by-line if it is
+  # too long to fit on one line. This will increase readability of
+  # debug output
```
I guess it would make sense to split the input command wherever we (likely) introduced manual concatenation via `&&` of several commands. That way it directly corresponds to the actual user input to the macro (unless one manually uses `&&` of course).

HaxScramper replied:
I don't think it is necessary to cram everything I want to do in one PR. If necessary I will open another one.

About splitting by `&&` vs spaces: it all comes down to use cases. Right now I mostly use with commands that require to pass a lot of different and long arguments (`latexmk` for example) together with absolute path for a file. Even if there is a single command it might take several lines in terminal. 

So it all comes down to use cases and it will requires some testing to get it look nice/

Vindaar commented on shell.nim:
```diff
@@ -320,30 +385,65 @@ macro shellVerbose*(cmds: untyped): untyped =
+    shellVerboseImpl `debugConfig`:
+      `cmds`
+
+
```
Extra space.

Vindaar commented on shell.nim:
```diff
@@ -320,30 +385,65 @@ macro shellVerbose*(cmds: untyped): untyped =
+      res.code = code
+      res
+
+
```
Same here.

Vindaar commented on shell.nim:
```diff
@@ -409,3 +509,15 @@ macro shellAssign*(cmd: untyped): untyped =
     echo result.repr
+
+when isMainModule:
+  var myConfig: set[DebugOutputKind] = {dokError}
```
What's the point of `myConfig = {dokError}` if it's not used in the next line and modified to `myConig = {dokOutput}` a few lines below? I guess the next line is just missing the `myConfig` arg?

Also this should also be moved to the tests (or just a test file) imo.

Vindaar commented:
Awesome! Yes, having `shellVerboseErr` be its own thing is better. 

Aside from the minor points, I consider this ready to be merged. Do you consider it complete?

HaxScramper commented:
I fixed formatting and removed `isMainModule` - i initially placed it to do quick-test on your runtime configuration and just forgot to remove it. 

I also think that this PR is ready to be merged: pretty-printing command output is something I will do in a near future but not in this PR.

Vindaar commented:
Thanks again!
